tests/integration/test_charm.py

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Test slurmd charm against other SLURM charms in the latest/edge channel."""
+
+import asyncio
+import pytest
+
+from helpers import (
+    fetch_slurmd_deps,
+    fetch_slurmctld_deps,
+)
+
+from pathlib import Path
+from pytest_operator.plugin import OpsTest
+from tenacity import retry
+from tenacity.stop import stop_after_attempt
+from tenacity.wait import wait_exponential as wexp
+
+SERIES = ["focal"]
+SLURMD = "slurmd"
+SLURMDBD = "slurmdbd"
+SLURMCTLD = "slurmctld"
+UNIT_0 = f"{SLURMD}/0"
+
+
+@pytest.mark.abort_on_fail
+@pytest.mark.parametrize("series", SERIES)
+@pytest.mark.skip_if_deployed
+async def test_build_and_deploy(ops_test: OpsTest, series: str, slurmd_charm):
+    """Test that the slurmd charm can stabilize against slurmctld, slurmdbd and percona."""
+    res_slurmd = fetch_slurmd_deps()
+    res_slurmctld = fetch_slurmctld_deps()
+
+    # Fetch edge from charmhub for slurmctld, slurmdbd and percona and deploy
+    await asyncio.gather(
+        ops_test.model.deploy(
+            SLURMCTLD,
+            application_name=SLURMCTLD,
+            channel="edge",
+            num_units=1,
+            resources=res_slurmctld,
+            series=series,
+        ),
+        ops_test.model.deploy(
+            SLURMDBD,
+            application_name=SLURMDBD,
+            channel="edge",
+            num_units=1,
+            series=series,
+        ),
+        ops_test.model.deploy(
+            "percona-cluster",
+            application_name="mysql",
+            channel="edge",
+            num_units=1,
+            series="bionic",
+        ),
+    )
+
+    # Attach ETCD resource to the slurmctld controller
+    await ops_test.juju("attach-resource", SLURMCTLD, f"etcd={res_slurmctld['etcd']}")
+
+    # Add slurmdbd integration to slurmctld
+    await ops_test.model.relate(SLURMCTLD, SLURMDBD)
+
+    # Add mysql integration to slurmdbd
+    await ops_test.model.relate(SLURMDBD, "mysql")
+
+    # TODO: It's possible for slurmd to be stuck waiting for slurmctld despite slurmctld and slurmdbd
```

Comment:
I would change TODO to IMPORTANT because TODO implies that you still have work to do, but in reality you are calling out a code smell:

```suggestion
    # IMPORTANT: It's possible for slurmd to be stuck waiting for slurmctld despite slurmctld and slurmdbd
```

Reply:
I like that, will add with my batch of changes